src/snowflake/snowpark/profiler.py

```diff
@@ -0,0 +1,146 @@
+        """
+        self.registered_stored_procedures = stored_procedures
+        sql_statement = (
+            f"alter session set python_profiler_modules='{','.join(stored_procedures)}'"
+        )
+        self.session.sql(sql_statement).collect()
+
+    def set_targeted_stage(self, stage: str):
+        """
+        Set targeted stage for profiler output.
+
+        Note:
+            The stage name must be a fully qualified name.
+
+        Args:
+            stage: String of fully qualified name of targeted stage
+        """
+        validate_object_name(stage)
+        self.stage = stage
+        if (
+            len(self.session.sql(f"show stages like '{self.stage}'").collect()) == 0
+            and len(
+                self.session.sql(
+                    f"show stages like '{self.stage.split('.')[-1]}'"
+                ).collect()
+            )
+            == 0
+        ):
+            self.session.sql(
+                f"create temp stage if not exists {self.stage} FILE_FORMAT = (RECORD_DELIMITER = NONE FIELD_DELIMITER = NONE )"
+            ).collect()
+        sql_statement = f'alter session set PYTHON_PROFILER_TARGET_STAGE ="{stage}"'
+        self.session.sql(sql_statement).collect()
+
+    def set_active_profiler(self, active_profiler: str):
+        """
+        Set active profiler.
+
+        Note:
+            Active profiler must be either 'LINE' or 'MEMORY' (case-sensitive). Active profiler is 'LINE' by default.
+        Args:
+            active_profiler: String that represent active_profiler, must be either 'LINE' or 'MEMORY' (case-sensitive).
+
+        """
+        if active_profiler not in ["LINE", "MEMORY", ""]:
+            raise ValueError(
+                f"active_profiler expect 'LINE', 'MEMORY' or empty string '', got {active_profiler} instead"
+            )
+        self.active_profiler = active_profiler
+        sql_statement = f"alter session set ACTIVE_PYTHON_PROFILER = '{self.active_profiler.upper()}'"
+        self.session.sql(sql_statement).collect()
+
+    def disable(self):
+        """
+        Disable profiler.
+        """
+        self.active_profiler = ""
+        sql_statement = "alter session set ACTIVE_PYTHON_PROFILER = ''"
+        self.session.sql(sql_statement).collect()
+
+    def _is_sp_call(self, query):
+        return re.match(self.pattern, query, re.DOTALL) is not None
+
+    def _get_last_query_id(self):
+        for query in self.query_history.queries[::-1]:
+            if query.sql_text.upper().startswith("CALL") or self._is_sp_call(
+                query.sql_text
+            ):
+                return query.query_id
+        return None
+
+    def show(self) -> None:
+        """
+        Show the profiles of the last executed stored procedure.
+
+        Note:
+            This function must be called right after the execution of stored procedure you want to profile.
+        """
+        query_id = self._get_last_query_id()
```

Comment:
This is non-deterministic. A session will be thread-safe so multiple SPs may be called concurrently.